### PR TITLE
fix: vectorize issue 1977 warm-start rerun

### PR DESCRIPTION
**Pull request by ll7**
## Summary
- add a config-first #1977 SLURM launcher that hydrates the preserved #1108/#749 W&B artifact and reruns BC warm-start PPO without recollecting trajectories
- repair PPO fine-tuning so it can use vectorized envs, CPU policy updates, CUDA predictive foresight, configurable hyperparameters, and periodic checkpoints
- update #1977/#749 warm-start fine-tune configs to use 22 subprocess envs and checkpoint every 500k timesteps
- preserve the cancelled full-run outcome as compact evidence rather than treating local `output/` checkpoints as durable proof

## Result boundary
- SLURM smoke `12688` completed in `00:08:18`: hydrated preserved artifact, trained `45,056` timesteps at `277 fps`, wrote periodic checkpoints, and wrote final smoke checkpoint/manifest.
- Full repaired job `12689` / `gse-1977-bcppo` was cancelled after `14:47:53` per maintainer direction because the success signal stayed extremely low.
- Last logged production state: `5,361,664` timesteps, iteration `119`, fps `100`, success rate `0`.
- No final PPO checkpoint, final manifest, or policy-analysis comparison exists for `12689`.
- The local periodic checkpoints through about 5M steps are diagnostic-only and not durable promoted artifacts.

## Evidence preserved
- Compact manifest: `docs/context/evidence/issue_1977_bc_warm_start_cancelled_2026-06-02/artifact_manifest.md`
- Context update: `docs/context/issue_1961_bc_warm_start_recoverability.md`

## Validation
- `uv run python -m py_compile robot_sf/training/imitation_config.py scripts/training/train_ppo_with_pretrained_policy.py`
- `uv run ruff check robot_sf/training/imitation_config.py scripts/training/train_ppo_with_pretrained_policy.py tests/training/test_train_ppo_with_pretrained_policy.py tests/integration/test_ppo_pretraining_pipeline.py`
- `uv run pytest tests/training/test_train_ppo_with_pretrained_policy.py tests/integration/test_ppo_pretraining_pipeline.py -q` (`16 passed`)
- `uv run python scripts/training/train_ppo_with_pretrained_policy.py --config configs/training/ppo_imitation/ppo_finetune_issue_1977_bc_warm_start_rerun.yaml --dry-run`
- `bash -n scripts/dev/sbatch_issue1977_bc_warm_start_ppo.sh`
- `git diff --check origin/main...HEAD` after merging latest `origin/main`
- `git diff --check` for the final evidence update

## Decision
Do not resubmit the same 10M continuation shape. Future work should either redesign the learned-progress objective or open a narrower intermediate-checkpoint analysis issue.

Refs #1977

### CODE REVIEW

File: scripts/training/train_ppo_with_pretrained_policy.py [hunk 1]
```diff
@@ -377,6 +646,15 @@ def _resolve_optional_path(name: str) -> Path | None:
         scenario_id=raw.get("scenario_id"),
         env_overrides=dict(raw.get("env_overrides") or {}),
         env_factory_kwargs=dict(raw.get("env_factory_kwargs") or {}),
+        num_envs=_parse_num_envs(raw.get("num_envs")),
+        num_envs_reserve_cores=int(raw.get("num_envs_reserve_cores", 0)),
+        worker_mode=str(raw.get("worker_mode", "auto")),
+        device=str(raw.get("device", "auto")),
```
[the remark this answers is not in the document]
Comment (ll7) [reply scripts/training/train_ppo_with_pretrained_policy.py]: Fixed in 2aaac554. YAML loading now normalizes blank/null `worker_mode` and `device`, and `PPOFineTuningConfig.from_raw()` also normalizes direct `worker_mode=None`. Added regression coverage for both paths.

Validation: `scripts/dev/run_worktree_shared_venv.sh -- pytest tests/integration/test_ppo_pretraining_pipeline.py tests/training/test_train_ppo_with_pretrained_policy.py -q` -> 18 passed; ruff check/format, py_compile, `scripts/dev/ci_driver.sh lint`, `bash -n`, and dry-run config command all passed.

File: SLURM/Auxme/issue_1977_bc_warm_start_ppo.sl [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+echo "[issue1977] WANDB_ARTIFACT=${WANDB_ARTIFACT}"
+echo "[issue1977] PPO_CONFIG=${PPO_CONFIG}"
+echo "[issue1977] RUN_POLICY_ANALYSIS=${RUN_POLICY_ANALYSIS}"
+echo "[issue1977] USE_SRUN=${USE_SRUN}"
+echo "[issue1977] git_commit=$(git rev-parse HEAD)"
+echo "[issue1977] git_branch=$(git branch --show-current || true)"
+
+uv sync
+
+run_in_allocation uv run python - <<'PY'
```
[the remark this answers is not in the document]
Comment (ll7) [reply SLURM/Auxme/issue_1977_bc_warm_start_ppo.sl]: Fixed in 2aaac554. The W&B hydration heredoc now runs directly with `uv run python - <<PY` in the batch shell instead of through `run_in_allocation`/possible `srun` stdin handling.

Validation: `bash -n SLURM/Auxme/issue_1977_bc_warm_start_ppo.sl scripts/dev/sbatch_issue1977_bc_warm_start_ppo.sh` passed; targeted tests, ruff checks, py_compile, `scripts/dev/ci_driver.sh lint`, and the fine-tune dry-run command also passed.

**Comment by ll7** [comment]
Not merge-ready yet: #1977 closeout evidence is still pending despite green CI.

Required before merging:
1. Complete job `12689` or record a fail-closed rerun outcome.
2. Provide a durable final PPO checkpoint pointer/manifest; local `output/*` files are not enough.
3. Publish policy-analysis results comparing the rerun policy against baselines `27dbe5xu` and `b60iopxt`.
4. Update this PR body and #1977 with explicit pass/fail, metric deltas, and artifact links before closing the issue.

I am treating this as blocked on evidence completeness, not code/CI failure.

**Comment by ll7** [comment]
Stopped job `12689` per maintainer direction after the success signal stayed extremely low.

```yaml
slurm_state: cancelled_intermediate
job_id: 12689
job_name: gse-1977-bcppo
partition: a30
elapsed: 14:47:53
start: 2026-06-01T15:54:17+02:00
end: 2026-06-02T06:42:10+02:00
last_logged_total_timesteps: 5361664
last_logged_iteration: 119
last_logged_success_rate: 0
last_logged_fps: 100
last_logged_updates: 1180
```

Intermediate result boundary:
- The run is diagnostic/intermediate evidence only. Do not count it as completed #1977 evidence or as a successful BC warm-start PPO rerun.
- The local cleanup trap synced outputs to `output/slurm/issue1977-bc-warm-start-ppo-job-12689/`.
- Periodic checkpoints are present at roughly 500k intervals through `4,999,940` steps under `output/slurm/issue1977-bc-warm-start-ppo-job-12689/benchmarks/expert_policies/checkpoints/issue_1977_ppo_finetune_v10_warm_start_rerun/`.
- No final fine-tuned checkpoint/manifest or post-training policy-analysis result exists, because the job was intentionally cancelled before completion.
- These paths are local `output/` artifacts and remain non-durable until promoted or recorded through an explicit artifact pointer.

Next action: preserve/promote the intermediate checkpoints only if they are useful as negative/diagnostic evidence, then select a different SLURM candidate rather than spending more allocation on this run shape.
